spring-session/src/main/java/org/springframework/session/hazelcast/HazelcastSessionRepository.java

```diff
@@ -0,0 +1,245 @@
+ *         .setExtractor(PrincipalNameExtractor.class.getName());
+ *
+ * Config config = new Config();
+ *
+ * config.getMapConfig("spring:session:sessions")
+ *         .addMapAttributeConfig(attributeConfig)
+ *         .addMapIndexConfig(new MapIndexConfig(
+ *                 HazelcastSessionRepository.PRINCIPAL_NAME_ATTRIBUTE, false));
+ *
+ * Hazelcast.newHazelcastInstance(config);
+ * </pre>
+ *
+ * This implementation listen for events on the Hazelcast-backed SessionRepository and
+ * translate those events into the corresponding Spring Session events. Publish the Spring
+ * Session events with the given {@link ApplicationEventPublisher}.
+ *
+ * <ul>
+ * <li>entryAdded - {@link SessionCreatedEvent}</li>
+ * <li>entryEvicted - {@link SessionExpiredEvent}</li>
+ * <li>entryRemoved - {@link SessionDeletedEvent}</li>
+ * </ul>
+ *
+ * @author Vedran Pavic
+ * @author Tommy Ludwig
+ * @author Mark Anderson
+ * @since 1.3.0
+ */
+public class HazelcastSessionRepository implements
+		FindByIndexNameSessionRepository<MapSession>,
+		EntryAddedListener<String, MapSession>,
+		EntryEvictedListener<String, MapSession>,
+		EntryRemovedListener<String, MapSession> {
+
+	/**
+	 * The principal name custom attribute name.
+	 */
+	public static final String PRINCIPAL_NAME_ATTRIBUTE = "principalName";
+
+	private static final Log logger = LogFactory.getLog(HazelcastSessionRepository.class);
+
+	private final IMap<String, MapSession> sessions;
+
+	private ApplicationEventPublisher eventPublisher = new ApplicationEventPublisher() {
+
+		public void publishEvent(ApplicationEvent event) {
+		}
+
+		public void publishEvent(Object event) {
+		}
+
+	};
+
+	/**
+	 * If non-null, this value is used to override
+	 * {@link MapSession#setMaxInactiveIntervalInSeconds(int)}.
+	 */
+	private Integer defaultMaxInactiveInterval;
+
+	private String sessionListenerId;
+
+	public HazelcastSessionRepository(IMap<String, MapSession> sessions) {
+		Assert.notNull(sessions, "Sessions IMap must not be null");
+		this.sessions = sessions;
+	}
+
+	@PostConstruct
+	private void init() {
+		this.sessionListenerId = this.sessions.addEntryListener(this, true);
+	}
+
+	@PreDestroy
+	private void close() {
+		this.sessions.removeEntryListener(this.sessionListenerId);
+	}
+
+	/**
+	 * Sets the {@link ApplicationEventPublisher} that is used to publish
+	 * {@link SessionDestroyedEvent}. The default is to not publish a
+	 * {@link SessionDestroyedEvent}.
+	 *
+	 * @param applicationEventPublisher the {@link ApplicationEventPublisher} that is used
+	 * to publish {@link SessionDestroyedEvent}. Cannot be null.
+	 */
+	public void setApplicationEventPublisher(
+			ApplicationEventPublisher applicationEventPublisher) {
+		Assert.notNull(applicationEventPublisher,
+				"ApplicationEventPublisher cannot be null");
+		this.eventPublisher = applicationEventPublisher;
+	}
+
+	/**
+	 * Set the maximum inactive interval in seconds between requests before newly created
```

Comment:
Nit-pick: "newly created" seems to be adding emphasis but I'm not sure why. Any session that has been inactive for the set amount of time will be invalidated, right?